pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,291 @@
+	return e.eventChannel
+}
+
+// Start spawns a goroutine to relist periodically.
+func (e *EventedPLEG) Start() {
+	go wait.Forever(e.watchEventsChannel, 0)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) >= cap(e.eventChannel) {
+		return false, fmt.Errorf("EventedPLEG: pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+
+	timestamp := e.clock.Now()
+	metrics.PLEGLastSeen.Set(float64(timestamp.Unix()))
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+
+	e.processCRIEvents(containerEventsResponseCh)
+}
+
+func (e *EventedPLEG) processCRIEvents(containerEventsResponseCh chan *runtimeapi.ContainerEventResponse) {
+	for event := range containerEventsResponseCh {
+		e.updatePodStatus(event)
+		e.processCRIEvent(event)
+	}
+}
+
+func (e *EventedPLEG) updatePodStatus(event *runtimeapi.ContainerEventResponse) {
+
+	podID := types.UID(event.PodSandboxMetadata.Uid)
+	podName := event.PodSandboxMetadata.Name
+	podNamespace := event.PodSandboxMetadata.Namespace
+
+	timestamp := e.clock.Now()
+	status, err := e.runtime.GetPodStatus(podID, podName, podNamespace)
+
+	if err != nil {
+		// nolint:logcheck // Not using the result of klog.V inside the
+		// if branch is okay, we just use it to determine whether the
+		// additional "podStatus" key and its value should be added.
+		if klog.V(6).Enabled() {
+			klog.ErrorS(err, "Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName), "podStatus", status)
+		} else {
+			klog.ErrorS(err, "Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName))
+		}
+	} else {
+		if klogV := klog.V(6); klogV.Enabled() {
+			klogV.InfoS("Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName), "podStatus", status)
+		} else {
+			klog.V(4).InfoS("Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName))
+		}
+		// Preserve the pod IP across cache updates if the new IP is empty.
+		// When a pod is torn down, kubelet may race with PLEG and retrieve
+		// a pod status after network teardown, but the kubernetes API expects
+		// the completed pod's IP to be available after the pod is dead.
+		status.IPs = e.getPodIPs(podID, status)
+	}
+
+	e.updateRunningPodMetric(status)
```

Comment:
for this metric to be right, i think if we lose connection, we must re-establish a list to know the initial value this is delta against.

Reply:
We now attempt to re-connect on failure and force a relisting to start from the correct set of metric values - https://github.com/kubernetes/kubernetes/pull/111384/files#diff-465aae123f6877816ff0428968fc74f503fe15b27b434480bf57b79e62c7fa2fR120